File: royale/source/android/JRoyale/jroyale/src/main/java/com/pmdtec/jroyale/listener/ExposureListener.java
/*
 * Copyright (C) 2018 pmdtechnologies ag
 *
 * THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
 * KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
 * PARTICULAR PURPOSE.
 */

package com.pmdtec.jroyale.listener;

/**
 * Provides the listener interface for handling auto-exposure updates in royale.
 *
 * To be notified of changes to the exposure, for example to update a UI slider, an application
 * may implement this interface and register itself as a listener to the ICameraDevice.  If the
 * application merely wishes to use auto-exposure but does not need to know that the exposure
 * has changed, it is not necessary to implement this listener.
 *
 * The exposure will be changed for future captures, but there may be another capture before the
 * new values take effect.  An application that needs the values for a specific set of captured
 * frames should use the metadata provided as part of the capture callback, for example in
 * DepthData::exposureTimes.
 */
@FunctionalInterface
public interface ExposureListener
{
    /**
     * Will be called when the newly calculated exposure time deviates from currently set exposure time of the current UseCase.
     *
     * @param exposureTime Newly calculated exposure time
     * @param streamId Current stream identifier
     */
    void onNewExposure(long exposureTime, int streamId);
}

File: royale/source/RoyaleDotNet/ExposureListener.cs
/****************************************************************************\
* Copyright (C) 2015 Infineon Technologies
*
* THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
* KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
* IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
* PARTICULAR PURPOSE.
*
\****************************************************************************/

using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace RoyaleDotNet
{
    /// <summary>
    /// A compatibility wrapper around IExposureListener2, which should be used instead.
    ///
    /// Please note that the callback is only made when autoexposure is enabled; when refactoring
    /// code to use the new interface, consider the possibility that the code you're refactoring has
    /// never been called.
    /// </summary>
    // The RegisterExposureListener(IEL) function is marked obsolete, adding an obsolete attribute
    // here too only creates duplicate warnings.
    public interface IExposureListener
    {
        void OnNewExposure (UInt32[] exposureTimes);
    }

    //\cond HIDDEN_SYMBOLS
    internal class ExposureListener : IExposureListener2
    {
        internal IExposureListener listener;

        internal ExposureListener (IExposureListener l)
        {
            listener = l;
        }

        public void OnNewExposure (UInt32 exposureTime, UInt16 streamId)
        {
            UInt32[] times = {exposureTime};
            listener.OnNewExposure (times);
        }
    }
    //\endcond
}
